scripts/casefold_db.py

```diff
@@ -0,0 +1,258 @@
+from sydent.util import json_decoder
+from sydent.util.emailutils import sendEmail
+from sydent.util.hash import sha256_and_url_safe_base64
+from tests.utils import ResolvingMemoryReactorClock
+
+
+def calculate_lookup_hash(sydent, address):
+    cur = sydent.db.cursor()
+    pepper_result = cur.execute("SELECT lookup_pepper from hashing_metadata")
+    pepper = pepper_result.fetchone()[0]
+    combo = "%s %s %s" % (address, "email", pepper)
+    lookup_hash = sha256_and_url_safe_base64(combo)
+    return lookup_hash
+
+
+def update_local_associations(
+    sydent, db: sqlite3.Connection, send_email: bool, dry_run: bool
+):
+    """Update the DB table local_threepid_associations so that all stored
+    emails are casefolded, and any duplicate mxid's associated with the
+    given email are deleted.
+
+    :return: None
+    """
+    cur = db.cursor()
+
+    res = cur.execute(
+        "SELECT address, mxid FROM local_threepid_associations WHERE medium = 'email'"
+        "ORDER BY ts DESC"
+    )
+
+    # a dict that associates an email address with correspoinding mxids and lookup hashes
+    associations: Dict[str, List[Tuple[str, str, str]]] = {}
+
+    # iterate through selected associations, casefold email, rehash it, and add to
+    # associations dict
+    for address, mxid in res.fetchall():
+        casefold_address = address.casefold()
+
+        # rehash email since hashes are case-sensitive
+        lookup_hash = calculate_lookup_hash(sydent, casefold_address)
+
+        if casefold_address in associations:
+            associations[casefold_address].append((address, mxid, lookup_hash))
+        else:
+            associations[casefold_address] = [(address, mxid, lookup_hash)]
+
+    # list of arguments to update db with
+    db_update_args: List[Tuple[str, str, str, str]] = []
+
+    # list of mxids to delete
+    to_delete: List[Tuple[str]] = []
+
+    # list of mxids to send emails to letting them know the mxid has been deleted
+    mxids: List[Tuple[str, str]] = []
+
+    for casefold_address, assoc_tuples in associations.items():
+        db_update_args.append(
+            (
+                casefold_address,
+                assoc_tuples[0][2],
+                assoc_tuples[0][0],
+                assoc_tuples[0][1],
+            )
+        )
+
+        if len(assoc_tuples) > 1:
+            # Iterate over all associations except for the first one, since we've already
+            # processed it.
+            for address, mxid, _ in assoc_tuples[1:]:
+                to_delete.append((address,))
+                mxids.append((mxid, address))
+
+    # iterate through the mxids and send email, let's only send one email per mxid
+    if send_email and not dry_run:
+        for mxid, address in mxids:
+            processed_mxids = []
+
+            if mxid in processed_mxids:
+                continue
+            else:
+                templateFile = sydent.get_branded_template(
+                    "none",
```

Comment:
Sorry, I meant the Python value `None`. This will cause this condition to trigger: https://github.com/matrix-org/sydent/blob/main/sydent/sydent.py#L470-L472 - though ironically the lines just above it will set the brand to `None` if it's unknown, but it's better to set it to the right value straight away:

```suggestion
                    None,
```